File: java/Graphs/Kruskal.java
Algoritmo para hallar el arbol cobertor mínimo de un grafo  no dirigido y conexo. Utiliza la técnica de Union-Find(Conjuntos disjuntos) para detectar que aristas generan ciclos.
Requiere la clase Edge(con pesos).

static int v, e; //vertices, arcos
static int MAX=100005;     
static int parent[]= new int [MAX];
static ArrayList<Edge> edges;
static ArrayList<Edge> answer;

//UNION-FIND   
static int find(int i){
    parent[i] = ( parent[i] == i ) ? i : find(parent[i]);
    return parent[i];
}

static void unionFind(int x, int y){
    parent[ find(x) ] = find(y);
}

static void kruskall(){
    Edge actual;
    int aux, i, x,y;
    aux = i = 0;
    Collections.sort(edges);
        
    while( aux < (v-1) && i < edges.size() ){
        actual = edges.get(i);
        x = find(actual.src);
        y = find(actual.dest);

        if( x != y ){
            answer.add(actual);
            aux++;
            unionFind(x, y);
        }
        i++;
    }
}

File: c++/Graphs/Kruskal.cpp
Dado un grafo con pesos halla su arbol cobertor minimo. Debe agregarse Disjoint Set.

struct edge { int u, v, w; };

bool cmp(edge &a, edge &b) {
    return a.w < b.w;
}

const int MX = 1e5+5; // Cantidad maxima de nodos
vector<pii> g[MX]; // Lista de adyacencia
vector<edge> e; // Lista de aristas
int n, m; // Cantidad de nodos y aristas

void kruskall() {
    sort(e.begin(), e.end(), cmp);
    dsu ds(n);
    int cnt = 0;
    for (auto &ed : e) {
        if (ds.find(ed.u) != ds.find(ed.v)) {
            ds.unite(ed.u, ed.v);
            g[ed.u].pb({ed.v, ed.w});
            g[ed.v].pb({ed.u, ed.w});
            if (++cnt == n-1) break;
        }
    }
}

void init() {
    e.clear();
    for (int i = 0; i <= n; i++) {
        g[i].clear();
    }
}
